### PR TITLE
persist-txn: rewrite data_snapshot and data_listen_next unit tests

### PR DESCRIPTION
**Pull request by danhhz**
Using the example history I came up with to try to exercise all the interesting edge cases.

Touches https://github.com/MaterializeInc/materialize/issues/22173
Touches https://github.com/MaterializeInc/materialize/issues/22801

### Motivation

  * This PR adds a known-desirable feature.

### Tips for reviewer

I'd love to go over the answers in this test with the two of you before trying to fix up the impl to pass it :D. (I think all the disagreements with the current impl are places where we're being too conservative, plus "eager" doesn't use this.)

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/persist-txn/src/txn_cache.rs
Comment (danhhz) [inline src/persist-txn/src/txn_cache.rs:1064]: For register and forget we could reasonably answer _either_ of `EmitLogicalProgress(x)` or `ReadDataTo(x)` (at least after Joe's PR lands). We know that the upper will physically advance, but we also know that timestamp is empty for this shard

File: src/persist-txn/src/txn_cache.rs
Comment (danhhz) [inline src/persist-txn/src/txn_cache.rs:1097]: The current impl returns `ReadDataTo(8)`, which is also correct and definitely pessimistic.

File: src/persist-txn/src/txn_cache.rs
Comment (jkosh44) [inline src/persist-txn/src/txn_cache.rs:1046]: Just thinking out loud about why this is expected.

- The cache knows everything < 2.
- `d0` is not registered in the cache.
- We know the shard can't be written to via txn < 2.
- So go read the shard normally up to 2.

Does that sound right?

File: src/persist-txn/src/txn_cache.rs
Comment (jkosh44) [reply src/persist-txn/src/txn_cache.rs:1074]: What is `latest_write` supposed to mean? If it's latest unapplied write, then `Some(7)` sounds wrong if we know it's applied. Unless it's more along the lines of "everything <= this timestamp is definitely applied. everything > this timestamp is unknown", then I agree either `Some(7)` or `None` sound correct to me.

File: src/persist-txn/src/txn_cache.rs
Comment (jkosh44) [inline src/persist-txn/src/txn_cache.rs:1085]: I'm having a little bit of trouble figuring out the difference between all of these. We don't actually write to anything in any of them.

File: src/persist-txn/src/txn_cache.rs
Comment (jkosh44) [inline src/persist-txn/src/txn_cache.rs:1110]: What should `c.data_listen_next(&d0, 14)` return?

File: src/persist-txn/src/txn_cache.rs
Comment (danhhz) [reply src/persist-txn/src/txn_cache.rs:1085]: Backing up a step, what I'm trying to solve here:

- For a table deletion, we forget the shard at some timestamp, removing it from the txns set
- We then advance the upper of the shard to empty antichain (this one of the necessary conditions to "delete" a shard)
- So, I would sort of expect that the txns_operator needs to continue to make progress even for shards that are no longer registered to the txns set.

- Similarly, because of the migration story, a shard might have been written to before it was registered. We probably want to ensure the txns_progress operator handles this situation correctly as well

- Finally, the unit test that registers and forgets stuff while doing direct and txns writes shook out a lot of bugs, so seems nice to support arbitrary interleavings here just so we can keep that test around

So, then, what are the expected semantics of txns_progress? I think my answer is that at any point it needs to be able to advance its own progress to`max(physical data shard upper, latest ts that the shard is registered in txns)`, the latter tracking the txns shard upper when the shard is registered.

Circling back to your question, this specific case tests that the shard is unregistered, it is not written to, and the txns shard advances. Basically, what happens is that we read what got written to the physical data shard at 11 (by reading to exclusive 12). If the next event in the data shard's timeline is a direct write, this read will resolve whenever that direct write happens. If the next event in the data shard's timeline is a register, then your PR will cause the data shard's upper to physically advance at that point, resolving that read.

File: src/persist-txn/src/txn_cache.rs
Comment (danhhz) [reply src/persist-txn/src/txn_cache.rs:1110]: WaitForTxnsProgress! (I'll add this in the next round of updates)

File: src/persist-txn/src/txn_cache.rs
Comment (jkosh44) [reply src/persist-txn/src/txn_cache.rs:1085]: So the comments like `not written` and `written directly` are just hypothetical. From our point of view we just care that from `[9, 12)` we know that the data shard is not registered with the txn shard so readers need to consult the shard directly?

Comment (jkosh44) [reply src/persist-txn/src/txn_cache.rs:1085]: > Curious if there's something I could write here that would have made it more obvious when you read it the first time?

Honestly, I'm not sure. It's probably fine as is.

File: src/persist-txn/src/txn_cache.rs
Comment (aljoscha) [inline src/persist-txn/src/txn_cache.rs:1109]: why is it interesting that we have to txn writes back-to-back? I forgot why we wanted that :see_no_evil:

File: src/persist-txn/src/txn_cache.rs
Comment (jkosh44) [inline src/persist-txn/src/txn_cache.rs:1165]: Should this be `assert_eq!(c.data_listen_next(&d0, 14), WaitForTxnsProgress);`?

File: src/persist-txn/src/txn_cache.rs
Comment (danhhz) [reply src/persist-txn/src/txn_cache.rs:1130]: Good question. Seems like ideally it'd at least have one of the 2s be an 8. Less clear to me if we want to make both of them be 8. On one hand, it'd be nice to test our actual behavior. On the other, it feels a bit weird to be hardcoding the retraction bug in the test

File: src/persist-txn/src/txn_cache.rs
Comment (danhhz) [reply src/persist-txn/src/txn_cache.rs:1168]: Oh yup. I think this must have been me incorrectly doing the comment immediately above this https://github.com/MaterializeInc/materialize/pull/23911/files#r1428214117

**Review by danhhz — COMMENTED** [review]
Started, but didn't finish, responding to your questions

File: src/persist-txn/src/txn_cache.rs
Comment (jkosh44) [reply src/persist-txn/src/txn_cache.rs:1150]: Ok, I'm not actually sure about `EmitLogicalProgress(14)` anymore. I originally said that because of the following test case above:
```
        // ts 8 (forget). Forget guarantees that all writes are applied, so no
        // latest_write. It also currently is inclusive, so we can emit logical
        // progress, though we might want to make the interval half-open?
        c.push_register(d0, 2, -1, 2);
        testcase(&mut c, 8, d0, ds(None, 8, 9), EmitLogicalProgress(9));
```

However, I think I have a larger question abou how `data_snapshot` should treat forgets. Imagine that I have the following series of events for some data shard `d0`:


```

        as_of_1            as_of_2
    -------|--------|---------|--------|--------
                  write              forget
```

What should `data_listen_next(d0, as_of_1)`, `data_listen_next(d0, write)`, `data_listen_next(d0, as_of_2)` return? One option that currently makes the most sense to me is the following

- `data_listen_next(d0, as_of_1)` = `ReadToData(write)`
- `data_listen_next(d0, write)` = `EmitLogicalProgress(forget)`
- `data_listen_next(d0, as_of_2)` = `EmitLogicalProgress(forget)`

This rationale here is that we wait for all writes to be applied, but don't have to wait for forgets to be applied.

Another option is the following

- `data_listen_next(d0, as_of_1)` = `ReadToData(forget)`
- `data_listen_next(d0, write)` = `ReadToData(forget)`
- `data_listen_next(d0, as_of_2)` = `ReadToData(forget)`

The rationale here is that we always wait for both writes and forgets to be applied and it ends up potentially involving less chatter with `data_listen_next`.

Yet another option, which *I think* is what the test is testing for is the following:

- `data_listen_next(d0, as_of_1)` = `ReadToData(forget)`
- `data_listen_next(d0, write)` = `EmitLogicalProgress(forget)`
- `data_listen_next(d0, as_of_2)` = `EmitLogicalProgress(forget)`

The rationale here is that we always wait for writes to be applied, but we only wait for forgets to be applied if there's a write inbetween `as_of` and the forget timestamp. This also involves less chatter with `data_listen_next` than the first approach, but makes the logic for `forget` more complicated.

Comment (jkosh44) [reply src/persist-txn/src/txn_cache.rs:1150]: Those examples probably are full of off-by-one errors.

File: src/persist-txn/src/txn_cache.rs
Comment (jkosh44) [reply src/persist-txn/src/txn_cache.rs:1154]: This is also related to https://github.com/MaterializeInc/materialize/pull/23911/files#r1566264250

**Comment by jkosh44** [comment]
@danhhz there's still some open questions, but in the meantime I've made some assumptions as we think about them. I pushed https://github.com/MaterializeInc/materialize/pull/23911/commits/3526bea4e59099d73ef0682e50fd1d9a98cc6dd0 which gets the test passing and hopefully doesn't break anything else.

File: src/persist-txn/src/txn_cache.rs
Comment (jkosh44) [reply src/persist-txn/src/txn_cache.rs]: Ok, I've been thinking about this some more. I think for now both should be `8`. There's an invariant that `ts <= compact_ts` so if we change `ts` to `8`, we'll also need to change `compact_ts` to `8`. I think `ts` should be `8` because we definitely need to encode the timestamp of the forget itself to be able to reason about when the forget happened. How we encode the corresponding register timestamp is still an open question.

I've left a comment that includes a link to the forget consolidation issue so that when we fix that issue Nightly will start failing with invalid GH references.

File: src/persist-txn/src/txn_cache.rs
Comment (danhhz) [inline src/persist-txn/src/txn_cache.rs:284]: I think this TODO is done? Mind removing the else locally and checking the the unit test fails?

File: src/persist-txn/src/txn_cache.rs
Comment (danhhz) [reply src/persist-txn/src/txn_cache.rs]: Makes sense to me!

File: src/persist-txn/src/txn_cache.rs
Comment (danhhz) [reply src/persist-txn/src/txn_cache.rs]: Great examples!

We talked about this offline, but I think I'm pretty flexible on what the answer is here. Personally, I wouldn't worry _too_ much about `data_listen_next` chatter, at least compared to factors like how easy it is to reason about and impl complexity. The policy you have here lgtm!

File: src/persist-txn/src/txn_cache.rs
Comment (danhhz) [inline src/persist-txn/src/txn_cache.rs:1195]: Wait, isn't this wrong? The shard was written to directly at `10`, and this would essentially lose that write, no?

Comment (jkosh44) [reply src/persist-txn/src/txn_cache.rs:1195]: Oops, yes great point. This should be `EmitLogicalProgress(9)` and then we should be returning a bunch of `ReadDataTo(12)`.

Comment (danhhz) [reply src/persist-txn/src/txn_cache.rs:1195]: Pretty worrisome that nothing in CI, including nightlies, seemed to catch this?

File: src/persist-txn/src/txn_cache.rs
Comment (jkosh44) [reply src/persist-txn/src/txn_cache.rs:284]: The reason is that it falls through to the final 

```
        // Nope, all caught up, we have to wait.
        WaitForTxnsProgress
```

which returns the right thing.

File: src/persist-txn/src/txn_cache.rs
Comment (jkosh44) [reply src/persist-txn/src/txn_cache.rs]: I'm not sure, is this tested anywhere outside of this unit test?

Comment (jkosh44) [reply src/persist-txn/src/txn_cache.rs]: I think in practice we're not forgetting and re-registering many shards and we aren't asking about times in the past with `data_listen_next` which is why this doesn't blow up anywhere.

File: src/persist-txn/src/txn_cache.rs
Comment (jkosh44) [reply src/persist-txn/src/txn_cache.rs]: I had a slight change of heart on which approach to take. It turns out that the implementation is A LOT simpler if we just look at the most recent register, write, forget and ignore anything inbetween the timestamp and those events.

File: src/persist-txn/src/txn_cache.rs
Comment (jkosh44) [reply src/persist-txn/src/txn_cache.rs]: My most recent commit should fix this though.

File: src/persist-txn/src/txn_cache.rs
Comment (jkosh44) [reply src/persist-txn/src/txn_cache.rs:289]: I'll remove the comment, but I actually like keeping the branch. It allows us to make the assumption below that the shard IS registered which simplifies the reasoning a bit.

Comment (danhhz) [reply src/persist-txn/src/txn_cache.rs:289]: Yeah, happy keeping the logic, but in that case, it does seem like it'd be nice to have something in the test that fails without it

Comment (jkosh44) [reply src/persist-txn/src/txn_cache.rs]: I think it's not possible to have a test that fails without this branch since it is strictly redundant with the final return. However, I think it's actually fine to just delete this branch too since we have asserts below that check that the shard is currently registered. I just pushed a commit to delete this branch.

File: src/persist-txn/src/txn_cache.rs
Comment (jkosh44) [reply src/persist-txn/src/txn_cache.rs]: FWIW here's a repro of this test case:

Note timestamp `t_n` is ordered by the value of `n`.

1. Register shard at `t_1`.
2. Forget shard at `t_2`.
3. Write directly to shard at `t_4`.
4. Register shard at `t_5`.
5. Call `data_listen_next` at `t_3`. This should return at least `ReadDataTo(t_4)` and not `EmitlogicalProgress(t_4)`.

Comment (danhhz) [reply src/persist-txn/src/txn_cache.rs]: Good news! `cargo stress -p mz-persist-txn --lib` got this failure at the commit with the bug (after 1929 attempts). Haven't checked yet what happens at the commit with the fix

```
thread 'txns::tests::stress_correctness' panicked at src/persist-txn/src/txns.rs:1439:21:
assertion `left == right` failed
  left: [("w0s12", 91, 1), ("w0s33", 91, 1), ("w0s36", 91, 1), ("w0s40", 91, 1), ("w0s45", 91, 1), ("w0s55", 91, 1), ("w0s59", 93, 1), ("w1s20", 91, 1), ("w1s39", 91, 1), ("w1s40", 91, 1), ("w1s51", 91, 1), ("w1s68", 91, 1), ("w1s71", 91, 1), ("w1s72", 91, 1)]
 right: [("w0s12", 91, 1), ("w0s33", 91, 1), ("w0s36", 91, 1), ("w0s40", 91, 1), ("w0s45", 91, 1), ("w0s55", 91, 1), ("w0s59", 93, 1), ("w1s20", 91, 1), ("w1s39", 91, 1), ("w1s40", 91, 1), ("w1s51", 91, 1), ("w1s68", 91, 1), ("w1s71", 91, 1), ("w1s72", 91, 1), ("w1s98", 107, 1)]
```

Comment (danhhz) [reply src/persist-txn/src/txn_cache.rs]: Yup, fixed at the latest commit!

```
20129 runs so far, 0 failures, over 3724s
```

**Review by jkosh44 — APPROVED** [review]

File: src/persist-txn/src/txn_cache.rs
Comment (jkosh44) [reply src/persist-txn/src/txn_cache.rs:285]: > > Specifically in this case, apply_le does guarantee that we eventually move the physical upper of the shard past the forget_ts (right? I just read the code again to be sure but worth confirmation).
>
> That's my understanding as well. I'll double check though.

Yes this is correct. 
- Commit that made this change: https://github.com/MaterializeInc/materialize/commit/1aff8ea19218f745862de20561537298ab77d27f
- Code that applies forgets: https://github.com/MaterializeInc/materialize/blob/8d9f10b5467990ef1fac28ea7b815a531b8946a4/src/persist-txn/src/txns.rs#L566-L578

Comment (jkosh44) [reply src/persist-txn/src/txn_cache.rs]: Just pushed these changes.

Comment (danhhz) [reply src/persist-txn/src/txn_cache.rs]: 😍 Okay I think this is all good to go! Gonna kick off one last nightlies run on this, but I think this is finally ready to merge. Thanks again for your patience on the back-and-forth here, I really like where this ended up

**Review by danhhz — COMMENTED** [review]
LGTM!

**Comment by jkosh44** [comment]
> Gonna kick off one last nightlies run on this

Looks like a lot of the nightlies failures are due to this branch not being rebased in awhile.

**Comment by danhhz** [comment]
Rebased! Kicking off a new nightlies

**Comment by danhhz** [comment]
Ooops I messed up that rebase. Redo!

**Comment by jkosh44** [comment]
Nightlies are all green except for a test that's failing on `main`. I'm going to merge.

**Pull request merged**
